renderdoc/driver/vulkan/vk_resources.cpp

```diff
@@ -3207,6 +3207,1292 @@ VkImageAspectFlags FormatImageAspects(VkFormat fmt)
+  for(auto src = begin(); src != end(); ++src)
+  {
+    arr.push_back(*src);
+  }
+}
+
+void ImageSubresourceMap::FromArray(const rdcarray<TaggedImageSubresourceState> &arr)
+{
+  if(arr.empty())
+  {
+    RDCERR("No values for ImageSubresourceMap");
+    return;
+  }
+  Split(arr.front().range);
+  if(m_values.size() != arr.size())
+  {
+    RDCERR("Incorrect number of values for ImageSubresourceMap");
+    return;
+  }
+  auto src = arr.begin();
+  auto dst = begin();
+  while(src != arr.end())
+  {
+    if(src->range != dst->range())
+      RDCERR("Subresource range mismatch in ImageSubresourceMap");
+    else
+      dst->SetState(src->state);
+    ++src;
+    ++dst;
+  }
+}
+
+void ImageSubresourceMap::FromImgRefs(const ImgRefs &imgRefs)
+{
+  bool splitLayers = imgRefs.areLayersSplit;
+  bool splitDepth = false;
+  if(GetImageInfo().extent.depth > 1)
+  {
+    RDCASSERT(GetImageInfo().layerCount == 1);
+    splitLayers = false;
+    splitDepth = true;
+  }
+  Split(imgRefs.areAspectsSplit, imgRefs.areLevelsSplit, splitLayers, splitDepth);
+  RDCASSERT(!(AreLayersSplit() && IsDepthSplit()));
+
+  for(auto dstIt = begin(); dstIt != end(); ++dstIt)
+  {
+    int aspectIndex = imgRefs.AspectIndex((VkImageAspectFlagBits)dstIt->range().aspectMask);
+    int level = (int)dstIt->range().baseMipLevel;
+    int layer = (int)(dstIt->range().baseArrayLayer + dstIt->range().baseDepthSlice);
+    dstIt->state().refType = imgRefs.SubresourceRef(aspectIndex, level, layer);
+  }
+}
+
+bool IntervalsOverlap(uint32_t base1, uint32_t count1, uint32_t base2, uint32_t count2)
+{
+  if((base1 + count1) < base1)
+  {
+    // integer overflow
+    if(count1 != VK_REMAINING_MIP_LEVELS)
+      RDCWARN("Integer overflow in interval: base=%u, count=%u", base1, count1);
+    count1 = UINT32_MAX - base1;
+  }
+  if((base2 + count2) < base2)
+  {
+    // integer overflow
+    if(count2 != VK_REMAINING_MIP_LEVELS)
+      RDCWARN("Integer overflow in interval: base=%u, count=%u", base2, count2);
+    count2 = UINT32_MAX - base2;
+  }
+  if(count1 == 0 || count2 == 0)
+    return false;    // one of the intervals is empty, so no overlap
+  if(base1 > base2)
+  {
+    std::swap(base1, base2);
+    std::swap(count1, count2);
+  }
+  return base2 < base1 + count1;
+}
+
+bool IntervalContainedIn(uint32_t base1, uint32_t count1, uint32_t base2, uint32_t count2)
+{
+  if((base1 + count1) < base1)
+  {
+    // integer overflow
+    if(count1 != VK_REMAINING_MIP_LEVELS)
+      RDCWARN("Integer overflow in interval: base=%u, count=%u", base1, count1);
+    count1 = UINT32_MAX - base1;
+  }
+  if((base2 + count2) < base2)
+  {
+    // integer overflow
+    if(count2 != VK_REMAINING_MIP_LEVELS)
+      RDCWARN("Integer overflow in interval: base=%u, count=%u", base2, count2);
+    count2 = UINT32_MAX - base2;
+  }
+  return base1 >= base2 && base1 + count1 <= base2 + count2;
+}
+
+bool ValidateLevelRange(uint32_t &baseMipLevel, uint32_t &levelCount, uint32_t imageLevelCount)
```

Comment:
Perhaps these functions should be renamed. They are validating the range, but also updating the range, both to handle `VK_REMAINING_*` values, and to try to prevent crashes if there is an incorrect range (possibly due to bad input from the application being captured).

Reply:
That makes sense, sorry I missed that it was updating the range at the same time as error checking.

I'm trying to think if there's a better name for that than validate and I'm not coming up with good options. Perhaps `Sanitise`? Similar to `SanitiseOldImageLayout`, though for the `VK_REMAINING_*` values it's not really sanitisation. It at least conveys more clearly that this is expected to be manipulating inputs into expected ranges, so that even if it does have some checks for incorrect ranges you don't make the mistake I did of thinking it's read-only debugging checks.

I think it's fine if you want to leave it as-is. Past the initial review it should be less likely for a reader to make that mistake.